ACM-16260 - create infrastructure for ARO-HCP clusters using ASO (#87)
stolostron/cluster-api-installer: doc/Create-resources-using-aso.md

# Preparing infrastructure for ARO HCP using ASO
We can initialize infrastructure using `az` CLI command like here: [Creating an HCP via Cluster Service](https://github.com/Azure/ARO-HCP/blob/main/cluster-service/cluster-creation.md)

> I this document we are demonstrationg ho to do the same can using `k8s` and ASO [Azure Service Operator v2](https://azure.github.io/azure-service-operator/)

Reviewer: I don't understand this

Author: Updated

## Prerequirements

> 1. We need to have a service principal and store it in sp.json

Reviewer: ```suggestion
1. Create azure account service principal and store it in json file sp.json as follow
```

```bash
AZURE_SUBSCRIPTION_ID=$(az account show --query id --output tsv)
if [ ! -f sp.json ] ; then
    let "randomIdentifier=$RANDOM*$RANDOM"
    servicePrincipalName="msdocs-sp-$randomIdentifier"
    roleName="Contributor"
    echo "Creating SP for RBAC with name $servicePrincipalName, with role $roleName and in scopes /subscriptions/$AZURE_SUBSCRIPTION_ID"
    az ad sp create-for-rbac --name $servicePrincipalName --role $roleName --scopes /subscriptions/$AZURE_SUBSCRIPTION_ID > sp.json
fi
```

> 2. We need to have certificate manager

Reviewer: I believe as pre-request helm should be installed.
```suggestion
2. Install certificate manager using helm chart
```

kind: ResourceGroup
metadata:
  name: mv-rg
  namespace: default
spec:
  location: eastus
---
# az network vnet create -n <vnet-name> -g <resource-group> --subnet-name <subnet-name>
apiVersion: network.azure.com/v1api20201101
kind: VirtualNetwork
metadata:
  name: mv-vnet
  namespace: default
spec:
  location: eastus
  owner:
    name: mv-rg
  addressSpace:
    addressPrefixes:
      - 10.100.0.0/15
---
# az network nsg create -n <nsg-name> -g <resource-group>
apiVersion: network.azure.com/v1api20201101
kind: NetworkSecurityGroup
metadata:
  name: mv-sg
  namespace: default
spec:
  location: eastus
  owner:
    name: mv-rg
---
# az network vnet create -n <vnet-name> -g <resource-group> --subnet-name <subnet-name>
# az network vnet subnet update -g <resource-group> -n <subnet-name> --vnet-name <vnet-name> --network-security-group <nsg-name>
apiVersion: network.azure.com/v1api20201101
kind: VirtualNetworksSubnet
metadata:
  name: mv-subnet
  namespace: default
spec:
  owner:
    name: mv-vnet
  addressPrefix: 10.100.76.0/24
  networkSecurityGroup: 
    reference:
      name: mv-sg
      group: network.azure.com
      kind: NetworkSecurityGroup
---

> # az identity create -n ${USER}-${CS_CLUSTER_NAME}-cp-control-plane-${OPERATORS_UAMIS_SUFFIX} -g <resource-group>

Reviewer: Add all the required az identities for both [control-plane](https://github.com/Azure/ARO-HCP/blob/main/cluster-service/cluster-creation.md?plain=1#L83)   and [data-plane](https://github.com/Azure/ARO-HCP/blob/main/cluster-service/cluster-creation.md?plain=1#L112C21-L112C31)

Author: I did a loop for:
```bash
USER=mv
CS_CLUSTER_NAME=cluster1
OPERATORS_UAMIS_SUFFIX=$(openssl rand -hex 3)
for IDENTITY_NAME in \
    ${USER}-${CS_CLUSTER_NAME}-cp-control-plane-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-cluster-api-azure-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-cloud-controller-manager-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-ingress-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-disk-csi-driver-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-file-csi-driver-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-image-registry-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-cloud-network-config-${OPERATORS_UAMIS_SUFFIX} \                                                                                                                                                           
    ${USER}-${CS_CLUSTER_NAME}-cp-kms-${OPERATORS_UAMIS_SUFFIX} \                                                                                                                                                                            
    \
    ${USER}-${CS_CLUSTER_NAME}-dp-disk-csi-driver-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-dp-image-registry-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-dp-file-csi-driver-${OPERATORS_UAMIS_SUFFIX} \
    \
    ${USER}-${CS_CLUSTER_NAME}-service-managed-identity-${OPERATORS_UAMIS_SUFFIX} \
; do
kubectl apply -f - <<EOF
# az identity create -n ${IDENTITY_NAME} -g <resource-group>
apiVersion: managedidentity.azure.com/v1api20230131
kind: UserAssignedIdentity
metadata:
  name: ${IDENTITY_NAME}
  namespace: default
spec:
  location: eastus
  owner:
    name: mv-rg
EOF
done
```

# Preparing Infrastructure for ARO HCP Using ASO

> We need to initialize the prerequisite infrastructure to create an ARO-HCP cluster:
> 
> * This can be done using the `az` CLI command, as shown here: [Creating an HCP via Cluster Service](https://github.com/Azure/ARO-HCP/blob/main/cluster-service/cluster-creation.md)  
> * In this document, we demonstrate how to achieve the same using `k8s` and [Azure Service Operator v2](https://azure.github.io/azure-service-operator/) with a declarative approach.

Reviewer: ```suggestion
The necessary infrastructure for deploying an ARO-HCP cluster can be provisioned using a declarative approach with [Azure Service Operator v2](https://azure.github.io/azure-service-operator/) as part of the Cluster API Provider for Azure. In the following steps, we will create the required Azure resources, including a Resource Group, Network Security Group, Virtual Network (VNet), Subnet, and a User Assigned Managed Identity.

```

## Prerequisites

We expect the following:

> * A `kind` cluster is being used (or any other Kubernetes cluster)  

Reviewer: ```suggestion
* Docker or Podman with `kind` cluster is being used OR Openshift cluster v4.18 or Later  
```

* The following tools are installed:  

>   * `az` CLI (or a `sp.json` file is already created)  

Reviewer: may add url to `az` command install

>   * `kubectl` – these are prerequisites for creating the ARO-HCP infrastructure using the declarative approach  

Reviewer: Better to mention `OC` instead of `kubectl` as this our redhat repo

Reviewer: add url for install `oc`

  * `helm` – also required for setting up the infrastructure using the declarative approach

> 1. Create azure account service principal and store it in json file sp.json as follow

Reviewer: I guess we should add more info on the azure account ARO-HCP  subscription requirements. May be just refer to a URL how create azure account with ARO-HCP subscription

Author: I added:
```md
* Ensure you have access to the RH Azure tenant:
  * **RH account**: You need to have a Red Hat account to access the Red Hat Azure tenant (`redhat0.onmicrosoft.com`) where personal DEV environments are created
  * **Subscription access**: You need access to the `ARO Hosted Control Planes (EA Subscription 1)` subscription in the Red Hat Azure tenant. Consult the [ARO HCP onboarding guide](https://docs.google.com/document/d/1KUZSLknIkSd6usFPe_OcEYWJyW6mFeotc2lIsLgE3JA/)
  * `az login` with your Red Hat account
```

```bash
AZURE_SUBSCRIPTION_ID=$(az account show --query id --output tsv)
if [ ! -f sp.json ] ; then
    let "randomIdentifier=$RANDOM*$RANDOM"
    servicePrincipalName="msdocs-sp-$randomIdentifier"
    roleName="Contributor"
    echo "Creating SP for RBAC with name $servicePrincipalName, with role $roleName and in scopes /subscriptions/$AZURE_SUBSCRIPTION_ID"
    az ad sp create-for-rbac --name $servicePrincipalName --role $roleName --scopes /subscriptions/$AZURE_SUBSCRIPTION_ID > sp.json
fi
```
2. Install cert-manager using a Helm chart
This is required for non-OCP clusters, as OCP clusters include a different built-in cert-manager.
```bash
helm repo add jetstack https://charts.jetstack.io --force-update
helm repo update
helm install cert-manager jetstack/cert-manager --namespace cert-manager --create-namespace --set crds.enabled=true
````

> 3. Install ASO2 using a Helm chart 

Reviewer: Install CAPZ using helm chart and add URL https://cluster-api.sigs.k8s.io/user/quick-start.html ref.  ASO come with CAPZ by default.

export AZURE_CLIENT_SECRET=$(jq -r .password sp.json)
helm repo add aso2 https://raw.githubusercontent.com/Azure/azure-service-operator/main/v2/charts
helm repo update
helm upgrade --install --devel aso2 aso2/azure-service-operator \
        --create-namespace --wait --timeout 2m \
        --namespace=azureserviceoperator-system \
        --set azureSubscriptionID=$AZURE_SUBSCRIPTION_ID \
        --set azureTenantID=$AZURE_TENANT_ID \
        --set azureClientID=$AZURE_CLIENT_ID \
        --set useWorkloadIdentityAuth=true \
        --set crdPattern='resources.azure.com/*;containerservice.azure.com/*;keyvault.azure.com/*;managedidentity.azure.com/*;eventhub.azure.com/*;network.azure.com/*;authorization.azure.com/*'
```
4. Deploy the AZURE secret:
```bash
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
 name: aso-credential
 namespace: default
stringData:
 AZURE_SUBSCRIPTION_ID: "$AZURE_SUBSCRIPTION_ID"
 AZURE_TENANT_ID: "$AZURE_TENANT_ID"
 AZURE_CLIENT_ID: "$AZURE_CLIENT_ID"
 AZURE_CLIENT_SECRET: "$AZURE_CLIENT_SECRET"
EOF
```
## Crating infrastructure for ARO-HCP clusters
5. Then we can start such a deplouyment:
```bash
kubectl -f - <<EOF

> # az group create --name <resource-group> --location <location>

Reviewer: No need to add the equivalent `az` command

apiVersion: resources.azure.com/v1api20200601
kind: ResourceGroup
metadata:

>   name: mv-rg

Reviewer: ```suggestion
  name: aro-hcp-resgroup
```

  namespace: default
spec:
  location: eastus
---

> # az network vnet create -n <vnet-name> -g <resource-group> --subnet-name <subnet-name>

Reviewer: same just remove the `az` command

apiVersion: network.azure.com/v1api20201101
kind: VirtualNetwork
metadata:

>   name: mv-vnet

Reviewer: ```suggestion
  name: aro-hcp-vnet
```

  namespace: default
spec:
  location: eastus
  owner:
    name: mv-rg
  addressSpace:
    addressPrefixes:
      - 10.100.0.0/15
---
# az network nsg create -n <nsg-name> -g <resource-group>
apiVersion: network.azure.com/v1api20201101
kind: NetworkSecurityGroup
metadata:

>   name: mv-sg

Reviewer: ```suggestion
   name: aro-hcp-nsg
```
and similar for all CR examples instead of `mv-` better `aro-hcp-`

      group: network.azure.com
      kind: NetworkSecurityGroup
EOF
USER=mv
CS_CLUSTER_NAME=cluster1
OPERATORS_UAMIS_SUFFIX=$(openssl rand -hex 3)
for IDENTITY_NAME in \
    ${USER}-${CS_CLUSTER_NAME}-cp-control-plane-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-cluster-api-azure-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-cloud-controller-manager-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-ingress-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-disk-csi-driver-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-file-csi-driver-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-image-registry-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-cloud-network-config-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-cp-kms-${OPERATORS_UAMIS_SUFFIX} \
    \
    ${USER}-${CS_CLUSTER_NAME}-dp-disk-csi-driver-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-dp-image-registry-${OPERATORS_UAMIS_SUFFIX} \
    ${USER}-${CS_CLUSTER_NAME}-dp-file-csi-driver-${OPERATORS_UAMIS_SUFFIX} \
    \
    ${USER}-${CS_CLUSTER_NAME}-service-managed-identity-${OPERATORS_UAMIS_SUFFIX} \
; do 

> kubectl apply -f - <<EOF

Reviewer: better to push the output of the for loop to a file ex; AroHcpUserAssignedIdentity.yaml then
 `oc apply -f AroHcpUserAssignedIdentity.yaml`
